forem/forem
# [Team Email Login] Refactor Authentication Provider Enable/Disable

## What type of PR is this? (check all applicable)


## IMPORTANT!
Re: the discussion [here](https://dev-internal.slack.com/archives/C7GE84DEJ/p1604506132027200), Team Email Login will refactor the `SiteConfig.allow_both_email_and_password` functionality to ensure that Email Login is on all the time. Those changes are NOT a part of this PR, and will be made after this PR is merged (for clarity's sake).

## Description
This PR refactors the enabling and disabling of the Authentication Providers, according to this [design mockup](https://user-images.githubusercontent.com/816402/96671965-f5909180-1317-11eb-990b-9bf4cabfa5cd.jpg).
I have left pertinent notes throughout my code changes, explaining my thought process.
**Allow me to say I'm quite proud of this refactor 😄  I learned a lot, and looking forward to learning even more from the QA process 🙏** 

## Related Tickets & Documents
closes https://github.com/forem/InternalProjectPlanning/issues/167
- ~I'm going to wait on the merge of Lisa's [Email Auth UX Fixes PR](https://github.com/forem/forem/pull/11169) before finalizing work on this PR; I will need to incorporate several of her updates~

## QA Instructions, Screenshots, Recordings
1. Navigate to `localhost:xxxx/admin/config` as a user with `:admin` (or `:super_admin`) and `:single_resource_admin` roles.
2. Go through the process of enabling and disabling auth providers. Here are the configurations I could think of:
- Can enable all or any combo of available Auth Providers, unless Invite-Only Mode is on.
- Can disable all available Auth Providers when Email Auth is on _(this is currently not possible; one Auth Provider must always be on)_
- Cannot disable an Auth Provider **if it is the sole enabled provider AND Email Auth is off**
- When Invite-Only Mode is enabled, all Auth Providers are disabled (and their buttons are disabled too)
3. Ensure that, based on Auth Provider setting changes, the expected auth options display on the login/create-account frontend pages

## Added tests?


## Added to documentation?


## [optional] Are there any post deployment tasks we need to perform?
No
